### PR TITLE
KAA-1423: Add Endpoint Specific Configuration Support.

### PR DESCRIPTION
**Pull request by v-chyzhevskyi**
Process of updating configuration fixed.

### CODE REVIEW

File: server/common/nosql/cassandra-dao/src/test/java/org/kaaproject/kaa/server/common/nosql/cassandra/dao/EndpointSpecificConfigurationCassandraDaoTest.java [hunk 1]
```diff
@@ -61,7 +60,7 @@ public void testFindByEndpointKeyHashAndConfigurationVersion() throws Exception
     Assert.assertNull(found4);
   }
 
-  @Test(expected = KaaOptimisticLockingFailureException.class)
+  @Test
```
Comment (vtkhir) [inline server/common/nosql/cassandra-dao/src/test/java/org/kaaproject/kaa/server/common/nosql/cassandra/dao/EndpointSpecificConfigurationCassandraDaoTest.java]: Why `KaaOptimisticLockingFailureException` is not expected anymore in this test?

File: server/common/nosql/cassandra-dao/src/main/java/org/kaaproject/kaa/server/common/nosql/cassandra/dao/EndpointSpecificConfigurationCassandraDao.java [hunk 2]
```diff
@@ -86,7 +86,16 @@ public CassandraEndpointSpecificConfiguration findByEndpointKeyHashAndConfigurat
...
   public EndpointSpecificConfiguration save(EndpointSpecificConfigurationDto dto) {
     LOG.debug("Saving endpoint specific configuration {}", dto);
-    EndpointSpecificConfiguration configuration = save(new CassandraEndpointSpecificConfiguration(dto));
+    CassandraEndpointSpecificConfiguration configuration =
+            findByEndpointKeyHashAndConfigurationVersion(dto.getEndpointKeyHash(), dto.getConfigurationSchemaVersion());
+    if (configuration != null) {
+      configuration.setConfiguration(dto.getConfiguration());
+      configuration.setConfigurationVersion(dto.getConfigurationSchemaVersion());
+      configuration.setEndpointKeyHash(getByteBuffer(dto.getEndpointKeyHash()));
+      save(configuration);
```
Comment (vtkhir) [inline server/common/nosql/cassandra-dao/src/main/java/org/kaaproject/kaa/server/common/nosql/cassandra/dao/EndpointSpecificConfigurationCassandraDao.java]: move a call of `save` method out of `if ` block

File: server/common/nosql/mongo-dao/src/main/java/org/kaaproject/kaa/server/common/nosql/mongo/dao/EndpointSpecificConfigurationMongoDao.java [hunk 3]
```diff
@@ -42,22 +42,31 @@ public void removeByEndpointKeyHashAndConfigurationVersion(byte[] endpointKeyHas
...
   public EndpointSpecificConfiguration save(EndpointSpecificConfigurationDto dto) {
     LOG.debug("Saving endpoint specific configuration {}", dto);
-    MongoEndpointSpecificConfiguration configuration = save(new MongoEndpointSpecificConfiguration(dto));
+    MongoEndpointSpecificConfiguration configuration =
+            (MongoEndpointSpecificConfiguration) findByEndpointKeyHashAndConfigurationVersion(dto.getEndpointKeyHash(), dto.getConfigurationSchemaVersion());
+    if (configuration != null) {
+      configuration.setConfiguration(dto.getConfiguration());
+      configuration.setConfigurationVersion(dto.getConfigurationSchemaVersion());
+      configuration.setEndpointKeyHash(dto.getEndpointKeyHash());
+      save(configuration);
```
Comment (vtkhir) [inline server/common/nosql/mongo-dao/src/main/java/org/kaaproject/kaa/server/common/nosql/mongo/dao/EndpointSpecificConfigurationMongoDao.java]: move a call of `save` method out of `if ` block

File: server/common/dto/src/main/java/org/kaaproject/kaa/common/dto/EndpointSpecificConfigurationDto.java [hunk 4]
```diff
@@ -26,19 +26,19 @@
   private byte[] endpointKeyHash;
   private Integer configurationSchemaVersion;
   private String configuration;
-  private Long version;
+  private Long specificConfigurationVersion;
```
Comment (vtkhir) [inline server/common/dto/src/main/java/org/kaaproject/kaa/common/dto/EndpointSpecificConfigurationDto.java]: Why do we need to rename `version` -> `specificConfigurationVersion` in DTO?

File: server/common/nosql/mongo-dao/src/test/java/org/kaaproject/kaa/server/common/nosql/mongo/dao/EndpointSpecificConfigurationMongoDaoTest.java [hunk 5]
```diff
@@ -72,7 +71,7 @@ public void testFindByEndpointKeyHashAndConfigurationVersion() throws Exception
     Assert.assertNull(found4);
   }
 
-  @Test(expected = KaaOptimisticLockingFailureException.class)
```
Comment (vtkhir) [inline server/common/nosql/mongo-dao/src/test/java/org/kaaproject/kaa/server/common/nosql/mongo/dao/EndpointSpecificConfigurationMongoDaoTest.java:75]: Why `KaaOptimisticLockingFailureException` is not expected anymore in this test?

**Review by vtkhir — CHANGES_REQUESTED** [review]

**Review by vtkhir — APPROVED** [review]

**Review by garniy — APPROVED** [review]

**Pull request merged**